Fix #1918: Send in page innerText to ads classifier

kylehickinson opened the pull request:
Instead of sending plain HTML, we now send in the document's `innerText`, which doesn't include any HTML tags, images, etc.
This allows the ads classifier to properly classify pages to their correct type

## Summary of Changes

This pull request fixes issue #1918 

## Submitter Checklist:

- [x] *Unit Tests* are updated to cover new or changed functionality
- [x] User-facing strings use `NSLocalizableString()`

## Test Plan:

- Enable rewards & ads
- Open Console.app and filter by BraveRewards _or_ build and run with debugger and in the console filter by BraveRewards. Log you are looking for looks like this: `[ads_impl.cc.750] Successfully classified page at https://www.law.com/ as law-law.`
- Visit law.com, verify page is classified correctly as `law-law`
- Visit imdb.com, verify page is classified correctly as `arts & entertainment-film`

## Reviewer Checklist:

- [x] Issues include necessary QA labels:
  - `QA/(Yes|No)`
  - `release-notes/(include|exclude)`
  - `bug` / `enhancement`
- [x] Necessary [security reviews](https://github.com/brave/security/issues/new/choose) have taken place.
- [x] Adequate unit test coverage exists to prevent regressions.
- [x] Adequate test plan exists for QA to validate (if applicable).
- [x] Issue is assigned to a milestone (should happen at merge time).

kylehickinson committed:
Fix #1918: Send in page innerText to ads classifier

Instead of sending plain HTML, we now send in the document's `innerText`, which doesn't include any HTML tags, images, etc.
This allows the ads classifier to properly classify pages to their correct type

iccub commented on Client/Frontend/Browser/BrowserViewController/BVC+Rewards.swift:
```diff
@@ -115,18 +115,41 @@ extension Tab {
+        webView.evaluateJavaScript("document.documentElement.outerHTML.toString()", completionHandler: { html, _ in
+            defer { group.leave() }
             guard let htmlString = html as? String else { return }
+            htmlBlob = htmlString
```
Can drop the guard as just use `htmlBlob = html as? String`
defer wouldn't be needed to then

kylehickinson replied:
The `html` isn't a `String` though, so either we guard like this or we just do `htmlBlob = html as? String`. I can swap it to the latter if you think that's better

iccub replied:
It's now just one line of code more, no?

iccub commented on Client/Frontend/Browser/BrowserViewController/BVC+Rewards.swift:
```diff
@@ -115,18 +115,41 @@ extension Tab {
+                defer { group.leave() }
+                guard var innerText = text as? String else { return }
+                for c in ["\'", "\"", "\\", "\n", "\r", "\t", "\0"] {
+                    innerText = innerText.replacingOccurrences(of: c, with: " ")
```
Thinking whether this can be a performance problem for large websites since we run this multiple times, perhaps a regular expression or other refactoring could be done here?
I'm fine with current solution for now, we can iterate on it later

kylehickinson replied:
I'll see if I can update this with a regex pattern that will match against all of these. This is just the simplest solution, if performance is a concern we could also throw it on a bg thread since we're already using dispatch groups

iccub commented on Client/Frontend/Browser/BrowserViewController/BVC+Rewards.swift:
```diff
@@ -115,18 +115,41 @@ extension Tab {
+            webView.evaluateJavaScript("document.body.innerText", completionHandler: { text, _ in
+                defer { group.leave() }
+                guard var innerText = text as? String else { return }
+                for c in ["\'", "\"", "\\", "\n", "\r", "\t", "\0"] {
```
Could add some documentation, I am not sure what are these all characters for :D and whether this is a full list

kylehickinson replied:
This is the list Terry gave me, I'll add a comment but these are a bunch of the escape characters that they don't want in the classifier 🤷‍♂️

iccub replied:
that's enough for a comment in code for me :)

iccub reviewed (changes requested):
In general I think that logic inside of `evaluateJavaScript("document.body.innerText"`
should be move to a String/url extension and add a unit test or two for it

kylehickinson commented:
> In general I think that logic inside of `evaluateJavaScript("document.body.innerText"`
> should be move to a String/url extension and add a unit test or two for it

I'm not sure what you mean, what would we unit test against? Local pages?

iccub commented:
my idea was to just pass a String of sample html body website like a random wikipedia/mozilla page and verify there's no special characters, words are trimmed and properly

what do you think about it

kylehickinson commented:
Added `String.words`, and added a few tests for it, let me know if you want more extensive tests. Also switched away from using regex/replacement stuff, just using `String`'s built in `enumerateSubstrings` with `byWords` option. Should work way better in general, strip everything, and work in multiple languages

iccub reviewed (approved):
Looks great, much cleaner now

iccub commented on BraveShared/Extensions/StringExtensions.swift:
```diff
@@ -14,4 +14,18 @@ extension String {
+        var words: [String] = []
+        enumerateSubstrings(
+            in: startIndex..<endIndex,
+            options: .byWords
```
This solves the previous special characters trim right? the `\, \\ \r` and others

kylehickinson replied:
Yes, it strips all punctuation that is not part of the word itself (i.e. an apostrophe), I could add more tests including them, one already has a `\r` in it

iccub replied:
It's ok, no need to test apple's built in method so hard, the pr is good to go

tmancey reviewed (approved):
LGTM++

iccub commented:
Need a rebase to force a new CI build, gh actions still can't be cancelled
